src/compiler/checker.ts

```diff
@@ -27322,7 +27328,30 @@ namespace ts {
 
             return getCandidateForOverloadFailure(node, candidates, args, !!candidatesOutArray);
 
-            function chooseOverload(candidates: Signature[], relation: ESMap<string, RelationComparisonResult>, signatureHelpTrailingComma = false) {
+            function addImplementationSuccessElaboration(failed: Signature, diagnostic : Diagnostic) {
+                const oldCandidatesForArgumentError = candidatesForArgumentError;
+                const oldCandidateForArgumentArityError = candidateForArgumentArityError;
+                const oldCandidateForTypeArgumentError = candidateForTypeArgumentError;
+
+                const declCount = length(failed.declaration?.symbol.declarations);
+                const isOverload = declCount > 1;
+                const implDecl = isOverload ? failed.declaration?.symbol.declarations[declCount - 1] : undefined;

```

Comment:
Instead of grabbing the last signature, consider just finding the first signature with a body, since we already try to harden against out-of-order implementation signatures in other places.

Also, I don't know if this will succeed, even in valid code.

```ts
class C {
    foo() {
        // ...
    }
}

interface C {
    foo(...args: string[]): void
}
```